integration-tests/tests/network/runner.rs

```diff
@@ -113,11 +116,16 @@ pub enum Action {
     // Send ping from `source` with `nonce` to `target`
     PingTo(usize, usize, usize),
     // Check for `source` received pings and pongs.
-    CheckPingPong(usize, Vec<(usize, usize)>, Vec<(usize, usize)>),
+    CheckPingPong(usize, Vec<(usize, usize, Option<usize>)>, Vec<(usize, usize, Option<usize>)>),
     // Send stop signal to some node.
     Stop(usize),
     // Wait time in milliseconds
     Wait(usize),
+    #[cfg(feature = "adversarial")]
+    SetOptions {
+        target: usize,
+        max_num_peers: Option<u64>,
+    },
```

Comment:
@bowenwang1996 @matklad 
I'm trying to write a code that is supposed to compile if either we are running test or adversarial flag is not set.
The following doesn't compile
```

#[cfg(any(test, feature = "adversarial"))]
pub struct SetAdvOptions {
    pub disable_edge_signature_verification: Option<bool>,
    pub disable_edge_propagation: Option<bool>,
    pub disable_edge_pruning: Option<bool>,
    pub set_max_peers: Option<u64>,
}
```

when running
```
cargo test --all-targets
```

```
no `SetAdvOptions` in `types`
```
Any ideas?

Reply:
@bowenwang1996 @matklad I see that we run `cargo test` without `--feature adversarial`. My tests depend on having `adversarial` flag enables, which adds extra functionality, like setting flags, to make the test work.

Due to the way `[cfg(test)]` works, it can't be used to mark function or struct, that should be compiled only while running a test. The existing tests we have in `integration-test/network/*.rs` really on using methods from `PeerManager` that are only used for for testing. That's not great, Ideally I would prefer to put all those methods behind a feature flag.

Reply:
I think that the best solution would be to add `adversarial` flag to CI. Currently `adversarial` provides additional test functionality, but it is only used by python tests.

Of course if we were to simply add `adversarial` to existing 2 test pipelines, you could have a code that work with, but doesn't with adversarial. So the best solution I see so far would be to add 2 extra pipelines:
```
cargo test --features adversarial,nightly_protocol,nightly_protocol_features
cargo test --features adversarial
```

Reply:
@bowenwang1996 @matklad What do you think?

Reply:
If we want to have an ability to write adversarial integrated tests in Rust, then we indeed need to add an extra CI pipeline here -- `#[cfg(test)]` only works within a single crate.

However, I want to ask *why* we didn't have a use for such tests before, and why we need to introduce them here. I have two concerns with this:

* First, adding an extra CI pipeline is sort-of a big deal. Granted, we can just dedicate more machines to CI and this shouldn't affect actual merge time, but, still, it increases the overall fragility of the system.
* Second, I think the original idea behind adversarial testing is that you want to spawn a network of nodes, *some* of which are adversarial and some are not, and you want to test that normal nodes are functioning correctly in the presence of adversaries. That is, it's important for adversarial tests that there are several neards running, and that we are testing the one with adversarial features disabled. 

  What we are doing here feels different -- this is a rust integrated test, so there's only one binary, and we are testing an adversarial nodes against itself. That is, to me it makes sense (abstractly), that we should add a new feature flag here, like "extra-test-knobs", and use that. Re-using existing "adversarial" feature for this seems accidental, adversarial feels like it should have a different semantics. 

>The existing tests we have in integration-test/network/*.rs realy on using methods from PeerManager that are only used for for testing. That's not great

I think this approach is actually fine. Sure, exposing extra knobs just for testing *feels* wrong, but I don't think there are actual problems there? I don't think we need to worry about someone accidentally using such a function wrongly, besides clearly labeling them as "for testing", preferably by hiding them in some `mod test_utils` or `trait TestExt`. Though, it might be good to try to minimize the amount of such tweaks we need for tests. 

Another plausible alternative is to use `pub(crate)` for those extra knobs, and move all tests to be unit-tests. Note that this requires that tests don't actually need `nearcore` or other such crates which are not already a dependency of the network.

Reply:
> I think the original idea behind adversarial testing is that you want to spawn a network of nodes, some of which are adversarial and some are not

I don't think so. I believe the idea is that some nodes "could be" adversarial, but it doesn't mean that we _necessarily_ have to have normal nodes in a test when adversarial ones are present.

Reply:
Also agree with what @matklad said. It makes more sense at this point to go with the existing approach -- exposing test-only structs. While this is not optimal, it is an existing issue and it is better to not pile new problems on top of that.

Reply:
-  I partially agree with @matklad. However, we currently use `adversarial` already in our python tests that run through `Nayduck`. So by adding extra pipeline for running tests will not make use system more fragile than it already is. It's possible to make code changes that will make `nayduck` fail, but pass CI. I would rather caught issues early while running CI, then spend hours a few days later, trying to figure out which PR broke them.

Cost of additional machines, shouldn't be an issue.
- I don't like the idea of having test methods in production. For purpose of `integration-test/network/*.rs` we introduced new data structures, that are only used by test, but are not read in production. I would prefer to not worry, that a node may crash, or that test data structures may grow in memory usage infinitely large.
There are also other reason why separation of test code is a good idea, it's much easier to read the code when doing PR, when you can guarantee that some branches of the code don't even get executed in producition.
My plan was to add 1-2 more data structures, that would allow us to make testing much easier, but I would rather not have them execute in production.
- Moving to `mod test_utils` would be ideal, but the existing code can't be easily moved there.

FYI tests in ` integration-test/network/*.rs ` spawn multiple nodes. We are actually testing interaction between them.
My intention was to move some of the tests from python to CI. That would make development easier. Test would also run faster. However, we don't have access to all `adversarial` features in CI.

Though, I agree with the naming of `adversarial`. Renaming `adversarial` to `test-features` seems appropriate, we use it for testing only. It's not indented for production.

Reply:
My proposal is to move all `integration-test/network/*.rs` to be under `adversarial` flag, and all methods they use in `networking`. @bowenwang1996 If that's something that would be accepted I prepare PR.

Reply:
I created an issue for that: https://github.com/near/nearcore/issues/4893

Reply:
I see. So you need what is already behind adversarial feature to run the tests. In that case, we should just make sure we can run the tests you write in CI. Otherwise they are even worse than tests run by nayduck -- they would not be run at all.

Reply:
I created separate PR for renaming `adversarial` to `test_features`. I see that we changed `pipeline.xml` to use `--features adversarial`. Is that enough?

Reply:
This code still looks very strange to me. You have an enum that is purely used for testing but you want to gate one specific variant of it instead of the whole thing. Could you explain what is the rationale for that?